### ISSUE TITLE
Expected type Float got Half in Loss function

### ISSUE THREAD

**Issue by pertschuk** [NVIDIA/apex#545]
This is my loss model forward:
```
    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor):
        rep_a, rep_b = self.model(sentence_features)

        output = cosine_similarity(rep_a.float(), rep_b.float()) # errors on this line unless I add the .float() cast
        loss_fct = nn.MSELoss()
        if labels is not None:
            loss = loss_fct(output, labels.view(-1))
            return loss
        else:
            return [rep_a, rep_b], output
```
Initialized with:
```
model, optimizer = amp.initialize(loss_model, optimizer, opt_level=fp16_opt_level)
```

**Comment by mcarilli** [comment]
torch.nn.functional.cosine_similarity should be [on the list](https://github.com/NVIDIA/apex/blob/656d14b0c9792a1bcdc255b473dc2d6145d026ff/apex/amp/lists/functional_overrides.py#L46) of functions that cast to FP32 anyway with O1.  How are you importing cosine_similarity?

**Comment by pertschuk** [comment]
```python
from torch.nn.functional import cosine_similarity
```

Also it should be noted this is an intermittent issue, it was working fine before, and I didn't change anything I can think of that directly pertains to apex?

The file is question is here if that's at all helpful. 
https://github.com/koursaros-ai/sentence-transformers/blob/master/sentence_transformers/losses/BiCosineSimilarityLoss.py

**Comment by mcarilli** [comment]
Can you try `print(cosine_similarity)` before and after `amp.initialize` to verify that it's actually been patched?

**Comment by pertschuk** [comment]
```python
from torch.nn.functional import cosine_similarity
print(cosine_similarity)
model, optimizer = amp.initialize(loss_model, optimizer, opt_level=fp16_opt_level)
print(cosine_similarity)
```

Output: 
```
<built-in method cosine_similarity of type object at 0x7f711c5d20a0>
Selected optimization level O1:  Insert automatic casts around Pytorch functions and Tensor methods.

Defaults for this optimization level are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic
Processing user overrides (additional kwargs that are not None)...
After processing overrides, optimization options are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic
<built-in method cosine_similarity of type object at 0x7f711c5d20a0>
```

**Comment by mcarilli** [comment]
hmm that's weird.  Maybe it's not patching onto the global namespace properly, although it should.  Can you also try
```
print(torch.nn.functional.cosine_similiarity)
```
before and after `amp.initialize`?  The function id should change.
